homeassistant/components/telegram_poll.py

```diff
@@ -0,0 +1,160 @@
+"""Recieve telegram bot messages.
+
+This uses async long polling.
+Configuration:
+
+```yaml
+telegram_poll:
+    api_key : your bot token
+    allowed_chat_ids :
+    - id1
+    - id2
+```
+"""
+
+import asyncio
+import logging
+import aiohttp
+
+from homeassistant.const import EVENT_HOMEASSISTANT_START, \
+    EVENT_HOMEASSISTANT_STOP, CONF_API_KEY
+from homeassistant.core import callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+import voluptuous as vol
+
+_LOGGER = logging.getLogger(__name__)
+
+EVENT_TELEGRAM_COMMAND = 'telegram.command'
+
+ATTR_COMMAND = 'command'
+ATTR_USER_ID = 'user_id'
+ATTR_ARGS = 'args'
+ATTR_FROM_FIRST = 'from_first'
+ATTR_FROM_LAST = 'from_last'
+
+DOMAIN = 'telegram_poll'
+CONF_ALLOWED_CHAT_IDS = 'allowed_chat_ids'
+
+REQUIREMENTS = ['python-telegram-bot==5.3.0']
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Required(CONF_ALLOWED_CHAT_IDS): vol.All(cv.ensure_list,
+                                                     [cv.positive_int]),
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Setup is called when Home Assistant is loading our component."""
+    import telegram
+
+    # instance the Telegram bot
+    bot = telegram.Bot(config[DOMAIN][CONF_API_KEY])
+    websession = async_get_clientsession(hass)
+    allowed_chat_ids = config[DOMAIN][CONF_ALLOWED_CHAT_IDS]
+    update_url = '{0}/getUpdates'.format(bot.base_url)
+    update_id = 0
+    # Flag which is checked while running the incoming loop.
+    running = True
+
+    @callback
+    def _start_bot(_event):
+        """Start the bot."""
+        hass.loop.create_task(check_incoming())
+
+    @callback
+    def _stop_bot(_event):
+        """Stop the bot."""
+        nonlocal running
+        running = False
```

Comment:
This will prevent Home Assistant from shutting down until there are enough updates. Instead, `hass.async_add_job` will return a task object that you can use to cancel the current task. It will then raise an `CancelledError` which you should catch and just stop the task.